### ISSUE TITLE
Missing optimization: Remove needless flattening

### ISSUE THREAD

**Issue by f0k** [Theano/Theano#5804]
Theano currently does not identify and remove a needless flattening operation:
```python
import theano
import theano.tensor as T
x = T.matrix('x')
y = x.flatten(2)  # useless reshape
theano.printing.debugprint(theano.function([x], y))
```
Output:
```
DeepCopyOp [id A] ''   3
 |Reshape{2} [id B] ''   2
   |x [id C]
   |MakeVector{dtype='int64'} [id D] ''   1
     |Shape_i{0} [id E] ''   0
     | |x [id C]
     |TensorConstant{-1} [id F]
```
This needlessly computes `x.reshape((x.shape[0], -1))` although `x` only has two dimensions anyway.

----

This has a real impact when trying to implement spatial softmax with spatial cross-entropy. With the current 2D softmax and 2D categorical crossentropy implementations, a spatial softmax requires some dimshuffle and reshape operations before and after the softmax. This hinders optimization:
```python
import theano
import theano.tensor as T
x = T.tensor4('x')
t = T.itensor3('t')  # an integer target per image and pixel
# simulate a softmax across the channel axis:
sm = T.nnet.softmax(x.dimshuffle(0,2,3,1).flatten(2))
sm = sm.reshape(x.dimshuffle(0,2,3,1).shape).dimshuffle(0,3,1,2)
# compute categorical cross-entropy across the channel axis:
ce = T.nnet.categorical_crossentropy(sm.dimshuffle(0,2,3,1).flatten(2), t.flatten())
# compile and print optimized graph
theano.printing.debugprint(theano.function([x, t], ce))
```
Output:
```
CrossentropyCategorical1Hot [id A] ''   7
 |Reshape{2} [id B] ''   6
 | |Softmax [id C] ''   5
 | | |Reshape{2} [id D] ''   4
 | |   |InplaceDimShuffle{0,2,3,1} [id E] ''   2
 | |   | |x [id F]
 | |   |MakeVector{dtype='int64'} [id G] ''   3
 | |     |Shape_i{0} [id H] ''   1
 | |     | |x [id F]
 | |     |TensorConstant{-1} [id I]
 | |MakeVector{dtype='int64'} [id G] ''   3
 |Reshape{1} [id J] ''   0
   |t [id K]
   |TensorConstant{(1,) of -1} [id L]
```
The two dimshuffles after the softmax are cancelled out, but note the superfluous `Reshape{2} [id B]` which reshapes the Softmax output to the shape it already had (`(x.shape[0], -1)`). The result is exactly the same if we explicitly use `(sm.shape[0], x.shape[2], x.shape[3], x.shape[1])` instead of `x.dimshuffle(0,2,3,1).shape` -- Theano infers `sm.shape[0]` to be `x.shape[0]`.

If we avoid reshaping the softmax output to a 4D tensor and back to 2D, it works:
```
CrossentropySoftmaxArgmax1HotWithBias.0 [id A] ''   7
 |Reshape{2} [id B] ''   4
 | |InplaceDimShuffle{0,2,3,1} [id C] ''   2
 | | |x [id D]
 | |MakeVector{dtype='int64'} [id E] ''   3
 |   |Shape_i{0} [id F] ''   1
 |   | |x [id D]
 |   |TensorConstant{-1} [id G]
 |Elemwise{second,no_inplace} [id H] ''   6
 | |Subtensor{int64} [id I] ''   5
 | | |Reshape{2} [id B] ''   4
 | | |Constant{0} [id J]
 | |TensorConstant{(1,) of 0.0} [id K]
 |Reshape{1} [id L] ''   0
   |t [id M]
   |TensorConstant{(1,) of -1} [id N]
```
But in a framework like Lasagne, it would be more natural to have a spatial softmax nonlinearity with 4D input and output, and a spatial cross-entropy objective accepting 4D inputs and targets (or 4D inputs and 3D targets), so we cannot easily avoid the redundant reshape.

**Comment by lamblin** [comment]
Yes, we could update the optimization for useless reshapes to check for the constant "-1" in at most 1 place, and consider it a match.

**Comment by f0k** [comment]
Actually, the useless reshape optimization has a TODO to also consider it a match if only one place doesn't match:
https://github.com/Theano/Theano/blob/dc83dda/theano/tensor/opt.py#L4423
This would be much easier to implement, but it could hide user errors by silently removing an invalid reshape. Checking for `-1` would be better.

**Comment by nouiz** [comment]
https://github.com/Theano/Theano/pull/5815
